lib/SILOptimizer/Utils/ConstantFolding.cpp

```diff
@@ -2026,8 +2026,28 @@ ConstantFolder::processWorkList() {
           if (r->use_empty())
             continue;
 
-          // Otherwise, do the RAUW.
-          User->getResult(Index)->replaceAllUsesWith(C);
+          // Make sure that our uses are not all uninteresting folder inserted
+          // uses. If so, continue.
+          if (llvm::all_of(r->getUses(), [&](Operand *op) {
+                return uninterestingFolderInsertedUses.count(op->getUser());
+              }))
+            continue;
+
+          // Otherwise, do the RAUW, inserting a compensating destroy if we are
+          // moving up the ownership lattice from .owned -> .none and will not
+          // be able to eliminate the underlying value due to it not being a
+          // single value instruction.
+          bool needsCompensatingDestroy =
+              isa<MultipleValueInstructionResult>(r) &&
+              r.getOwnershipKind() == ValueOwnershipKind::Owned &&
+              C.getOwnershipKind() == ValueOwnershipKind::None;
+          r->replaceAllUsesWith(C);
```

Comment:
I find it hard to reason about correctness of this RAUW w.r.t. ownership. If it were expressed this way I know it would be correct
- Do the RAUW
- If the replaced value is owned, check if it's trivially dead now
- If it's not trivially dead, insert the destroy

Reply:
Sure. I am going to do something sort of like this. One issue is that we do not know until we have processed all of the results if the overall multiple value instruction will /actually/ be dead. I am going to try restructuring the loop here to enable me to do that.